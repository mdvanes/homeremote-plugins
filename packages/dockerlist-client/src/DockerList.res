// {..} means we are handling a JS object with an unknown
// set of attributes
@module external styles: {..} = "./DockerList.module.css"

type request
type response

@new external makeXMLHttpRequest: unit => request = "XMLHttpRequest"
@send
external addEventListener: (request, string, unit => unit) => unit = "addEventListener"
@get external response: request => response = "response"
@send external open_: (request, string, string) => unit = "open"
@send external send: request => unit = "send"
@send external abort: request => unit = "abort"

@get external status: request => int = "status"

module DockerListMod = {
  type t = {
    address: string,
    total_balance: float,
  }

  @scope("JSON") @val
  external // external parseResponse: response => t = "parse"
  parseResponse: response => {"message": array<string>} = "parse"

  //  let query = (~address, ~onDone, ~onError, ()) => {
  let query = (~address) => {
    let request = makeXMLHttpRequest()

    // request->addEventListener("load", () => onDone(request->response->parseResponse))
    // request->addEventListener("error", () => onError(request->status))
    request->addEventListener("load", () => {
      let response = request->response->parseResponse
      Js.log(response["message"])
    })
    request->addEventListener("error", () => {
      Js.log("Error logging here")
    })

    request->open_("GET", "https://dog.ceo/api/breeds/image/random/" ++ address)
    request->send

    // () => request->abort
  }

  @react.component
  let make = (~count: int) => {
    let times = switch count {
    | 1 => "once"
    | 2 => "twice"
    | n => Belt.Int.toString(n) ++ " times"
    }
    let msg = "Click me " ++ times

    // Runs only once right after mounting the component
    React.useEffect0(() => {
      // Run effects
      query(~address="3")
      None // or Some(() => {})
    })

    <div
      style={ReactDOM.Style.make(
        // ~backgroundColor="white",
        ~padding="2px",
        ~borderRadius="2px",
        (),
      )}>
      {React.string("Docker List")}
      <button className={styles["root"]}> {msg->React.string} </button>
    </div>
  }
}

// Js.log("Hello, World!")
