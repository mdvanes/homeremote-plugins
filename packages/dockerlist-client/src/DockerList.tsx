import React, { useRef } from "react";
// import "./StreamPlayer.css";
// import Elm from "react-elm-components";
// import Audio from "./Elm/Audio.elm";

// const setupPorts = (containerRef) => (ports) => {
//     ports.setPlayPauseStatusPort.subscribe((newStatus) => {
//         // NOTE: findDomNode has been deprecated
//         const audioElem = containerRef.current.children[0].querySelector(
//             "audio"
//         );
//         // Wait to let the audio elem be updated with a new cachebusting timestamp in Audio.elm `Cmd.batch [ Task.perform UpdateTimestamp Time.now, Cmd.map MsgControls controlsCmds ]`
//         setTimeout(() => {
//             if (newStatus === "Play") {
//                 audioElem.play();
//             } else {
//                 audioElem.pause();
//             }
//         }, 50);
//     });
// };

const DockerList = ({ url }) => {
    const containerRef = useRef(null);
    return (
        <div ref={containerRef}>
            <div
                style={{
                    backgroundColor: "white",
                    padding: 2,
                    borderRadius: 2,
                }}
            >
                Docker List
            </div>

            {/* <Elm
                src={Audio.Elm.Elm.Audio}
                flags={{ url }}
                ports={setupPorts(containerRef)}
            /> */}
        </div>
    );
};

export default DockerList;
