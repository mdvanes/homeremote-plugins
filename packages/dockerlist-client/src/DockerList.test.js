import React, { Component } from "react";
import { render, screen, waitFor, wait } from "@testing-library/react";
import { DockerListMod } from "./DockerList.bs";
import { json } from "express";
// import "./DockerList.bs.js";

// eslint-disable-next-line @typescript-eslint/no-var-requires
// const { DockerListMod } = require("./DockerList.bs");

// TODO convert to typescript?
// TODO convert to bs-jest?

const DockerList = DockerListMod.make;

jest.mock("./DockerList.module.css", () => "");

const FOO = {
    status: "received",
    containers: [
        {
            Id:
                "2757091519a2e227b40e6fae4f92b7f4c8b7b189037bda0bc3152ce2be96af9d",
            Names: ["/determined_edison"],
            State: "exited",
            Status: "Exited (1) 5 months ago",
        },
        {
            Id:
                "42aebf279f8c95488fab905d788f3caffee6afcaad240fc4aca68106c7173bfe",
            Names: ["/tinycors"],
            State: "exited",
            Status: "Exited (2) 13 hours ago",
        },
        {
            Id:
                "c7f51601f9870180859a2feb981613b30e3fcd282830b398185aa4facff21be1",
            Names: ["/hello_world"],
            State: "exited",
            Status: "Exited (0) 17 hours ago",
        },
    ],
};

// function getDockerList(url, onError) {
//     var __x = fetch(url + "/api/dockerlist");
//     return handleResponse(__x.then(function (prim) {
//                     return prim.text();
//                   }), onError, "error in getDockerList");

// global.fetch = jest.fn().mockResolvedValue({
//     then: (fn) => {
//         // fn();
//         jest.fn().mockResolvedValue(JSON.stringify(FOO));
//     },
// });

global.fetch = jest.fn().mockResolvedValue({
    text: () => Promise.resolve(JSON.stringify(FOO)),
});

// global.fetch = jest.fn().mockResolvedValue({
//     text: () => jest.fn().mockResolvedValue(FOO),
// });

describe("DockerList client", () => {
    it("trivial test", () => {
        expect(true).toBeTruthy();
    });

    // it("trivial render", () => {
    //     render(<DockerList url={"some_url"} />);
    //     screen.debug();
    // });

    it("trivial render", async () => {
        render(<DockerList url={"some_url"} />);

        await waitFor(() => {
            screen.getByText("determined_edison");
        });
        screen.debug();
    });
});

// In this case use `require`, see below. import StreamPlayer from "./StreamPlayer";

// let mockEmit = () => {
//     /* no-op */
// };

// // Example of naive mock:
// // jest.mock("react-elm-components", () => "mock-react-elm-components");

// // Example of FC mock:
// // jest.mock("react-elm-components", () => (props) => {
// //     // Do not destructure `props.ports` to keep destinct from inner `ports` argument
// //     console.log("props.ports", props.ports);
// //     // ports();
// //     const foo = (fn) => {
// //         mockEmit = fn;
// //     };
// //     const ports = { setPlayPauseStatusPort: { subscribe: foo } };
// //     // triggerPorts =
// //     props.ports(ports);
// //     return <p data-testid="mock-react-elm-components" />;
// // });

// const playSpy = jest.fn();
// const pauseSpy = jest.fn();

// window.HTMLMediaElement.prototype.play = playSpy;
// window.HTMLMediaElement.prototype.pause = pauseSpy;

// class MockReactElmComponent extends Component {
//     render() {
//         // Do not destructure `props.ports` to keep destinct from inner `ports` argument
//         const foo = (fn) => {
//             mockEmit = fn;
//         };
//         const ports = { setPlayPauseStatusPort: { subscribe: foo } };
//         this.props.ports(ports);
//         return (
//             <p data-testid="mock-react-elm-components">
//                 <audio />
//             </p>
//         );
//     }
// }

// With ref:
// jest.mock("react-elm-components", () => MockReactElmComponent);

// eslint-disable-next-line @typescript-eslint/no-var-requires
// const StreamPlayer = require("./StreamPlayer").default;

// jest.mock("./Elm/Audio.elm", () => ({
//     Elm: {
//         Elm: {
//             Audio: "mock-elm-audio",
//         },
//     },
// }));
