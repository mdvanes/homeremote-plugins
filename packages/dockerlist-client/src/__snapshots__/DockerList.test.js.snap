// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DockerList client can stop running container 1`] = `
<button
  class="  "
>
  <h1>
    determined_edison
  </h1>
  <p>
    Up 2 days
  </p>
</button>
`;

exports[`DockerList client can stop running container 2`] = `
<div>
  <div>
    <div>
      <button
        class="  "
      >
        <h1>
          determined_edison
        </h1>
        <p>
          Up 2 days
        </p>
      </button>
      <dialog>
        <p>
          Do you want to stop determined_edison?
        </p>
        <div>
          <button>
            cancel
          </button>
          <button
            style="background-color: darkblue; color: white;"
          >
            OK
          </button>
        </div>
      </dialog>
      <button
        class=" "
      >
        <h1>
          tinycors
        </h1>
        <p>
          Exited (2) 13 hours ago
        </p>
      </button>
      <dialog>
        <p>
          Do you want to start tinycors?
        </p>
        <div>
          <button>
            cancel
          </button>
          <button
            style="background-color: darkblue; color: white;"
          >
            OK
          </button>
        </div>
      </dialog>
      <button
        class=" "
      >
        <h1>
          hello_world
        </h1>
        <p>
          Exited (0) 17 hours ago
        </p>
      </button>
      <dialog>
        <p>
          Do you want to start hello_world?
        </p>
        <div>
          <button>
            cancel
          </button>
          <button
            style="background-color: darkblue; color: white;"
          >
            OK
          </button>
        </div>
      </dialog>
    </div>
  </div>
</div>
`;
